paint/PAINT_ShapeTool.cpp

```diff
@@ -1,16 +1,51 @@
 #include "PAINT_pch.h"
 #include "PAINT_ShapeTool.h"
+#include "PAINT_Brush.h"
+#include "GFX_Rectangle.h"
 
 using namespace paint;
 
-ShapeTool::ShapeTool()
+ShapeTool::ShapeTool(SDL_Renderer* renderer, SDL_Texture* texture)
 	: activeShape_(nullptr)
+	, renderer_(renderer)
+	, texture_(texture)
 {
-	
+	activeBrush_ = std::make_shared<Brush>(0);
 }
 
-void ShapeTool::setActiveShape(Shape* shape)
+void ShapeTool::setActiveShape(std::shared_ptr<Shape> shape)
 {
-	activeShape_ = shape;
+	activeShape_ = std::move(shape);
 }
 
+void ShapeTool::toolFunction(win::Coords mouseCoords, win::Coords prevMouseCoords, win::Coords startCoords, gfx::Rectangle refRect)
+{
+	clearLines();
+	const win::Coords rel = { mouseCoords.x - refRect.x, mouseCoords.y - refRect.y };
+	if (rel.x < 0)
+	{
+		mouseCoords.x = refRect.x;
+	}
+	if (rel.y < 0)
+	{
+		mouseCoords.y = refRect.y;
+	}
+	if (rel.y > refRect.height)
+	{
+		mouseCoords.y = refRect.y + refRect.height;
+	}
+	lines_ = activeShape_->shapeGeometry(startCoords, mouseCoords);
+	renderLines(renderer_, lines_);
+}
+
+void ShapeTool::toolFunctionEnd(win::Coords mouseCoords, win::Coords prevMouseCoords, win::Coords startCoords, gfx::Rectangle refRect)
+{
+	std::vector<gfx::Line> shiftedLines;
+	for (auto line : lines_)
+	{
+		shiftedLines.push_back({ line.x1 - refRect.x, line.y1 - refRect.y, line.x2 - refRect.x, line.y2 - refRect.y });
+	}
+	SDL_SetRenderTarget(renderer_, texture_);
```

Comment:
Just as a heads-up. It's probably the case that changing a render target is a fairly expensive operation. The fewer times it's changed, the quicker stuff runs. It's entirely possible that SDL has a "lazy" policy - it will only actually do the work if the render target is different from the current one, But we don't know that. So

(1) In the shiny renderer class we can cache the current render context and only call SDL_SetRenderTarget() if the pointers differ.
(2) Don't bother setting it to null in this block of code and ensure that every class that can render sets the correct value before doing its draw.
**COULD**

Reply:
Zoe is going to do this

Reply:
I've talked to Amy and we decided I can do this, as I've been the one handling the gfx::Renderer class so far.

Reply:
Done as requested, but the changes won't be in this branch. Now each rendering function requires an enum defining the render target, and if it is different to the previous one then it will change to the new target, but if it is the same it will not change.